### PR TITLE
state: bulk sql for unused links deletion

### PR DESCRIPTION
**Pull request by pared**
* [ ] Have you followed the guidelines in our
      [Contributing document](https://dvc.org/doc/user-guide/contributing)?

* [ ] Does your PR affect documented changes or does it add new functionality
      that should be documented? If yes, have you created a PR for
      [dvc.org](https://github.com/iterative/dvc.org) documenting it or at
      least opened an issue for it? If so, please add a link to it.

-----
Related to #2093

### CODE REVIEW

File: dvc/state.py [hunk 1]
```diff
@@ -449,6 +449,7 @@ def remove_unused_links(self, used):
                 remove(path)
                 unused.append(relpath)
 
-        for relpath in unused:
-            cmd = "DELETE FROM {} WHERE path = ?".format(self.LINK_STATE_TABLE)
-            self._execute(cmd, (relpath,))
+        cmd = "DELETE FROM {} WHERE path IN ({})".format(
```
Comment (efiop) [inline dvc/state.py]: And what if we reach the max length of the command?

File: dvc/utils/__init__.py [hunk 2]
```diff
@@ -382,3 +382,8 @@ def relpath(path, start=os.curdir):
     ):
         return path
     return os.path.relpath(path, start)
+
+
+def chunk(iterable, chunk_size=100):
```
Comment (efiop) [inline dvc/utils/__init__.py]: Isn't `to_chunks` the same thing?

Comment (efiop) [reply dvc/utils/__init__.py]: Also, where did 100 come from? Did you estimate it for sql or something?

Comment (pared) [reply dvc/utils/__init__.py]: not exactly, in to_chunks we define number of chunks, so chunk length varies depending on length of argument, in this case I am focusing on particular chunk length. As to `chunk_size`  the only criterion taken by me into account was maximum variables number, which, according to [docs](https://www.sqlite.org/limits.html) is 999. Do you think I should perform some performance tests before deciding to default?

Comment (pared) [reply dvc/utils/__init__.py]: For sure it should not be default in method. Its wrong to decide generic function default argument basing on particular use case scenario.

File: dvc/state.py [hunk 3]
```diff
@@ -449,6 +449,8 @@ def remove_unused_links(self, used):
                 remove(path)
                 unused.append(relpath)
 
-        for relpath in unused:
-            cmd = "DELETE FROM {} WHERE path = ?".format(self.LINK_STATE_TABLE)
-            self._execute(cmd, (relpath,))
+        for chunk_unused in chunk(unused, 100):
```
Comment (efiop) [inline dvc/state.py]: So why 100 then if max is 999? Also, don't hardcode these, maybe we could reference SQLITE_MAX_VARIABLE_NUMBER somehow?

File: dvc/utils/__init__.py (hunk 2 above)
Comment (efiop) [reply dvc/utils/__init__.py]: > not exactly, in to_chunks we define number of chunks, so chunk length varies depending on length of argument, in this case I am focusing on particular chunk length

Got it. :+1: 

>  As to chunk_size the only criterion taken by me into account was maximum variables number, which, according to docs is 999.

Left a comment about it down below.

> Do you think I should perform some performance tests before deciding to default?

Not really, the more - the better in this case, isn't it the point of this whole PR ? :slightly_smiling_face: Let's try to squeeze as much as we can.

File: dvc/state.py (hunk 3 above)
Comment (pared) [reply dvc/state.py]: I am unable to find this constant/ anything that looks like this constant inside sqlite package

Comment (pared) [reply dvc/state.py]: I updated with 999

File: dvc/utils/__init__.py [hunk 4]
```diff
@@ -382,3 +382,8 @@ def relpath(path, start=os.curdir):
     ):
         return path
     return os.path.relpath(path, start)
+
+
+def chunk(iterable, chunk_size):
```
Comment (efiop) [inline dvc/utils/__init__.py]: Ok, how about making `to_chunks` accept both desired number of chunks or chunk size? I really don't like having both `to_chunks` and `chunk` at the same time.

Comment (pared) [reply dvc/utils/__init__.py]: I can work with that :)

File: dvc/state.py [hunk 5]
```diff
@@ -7,11 +7,12 @@
 import logging
 
 from dvc.config import Config
-from dvc.utils import remove, current_timestamp, relpath
+from dvc.utils import remove, current_timestamp, relpath, chunk
 from dvc.exceptions import DvcException
 from dvc.utils.fs import get_mtime_and_size, get_inode
 from dvc.utils.compat import fspath_py35
 
+SQLITE_MAX_VARIABLES_NUMBER = 999
```
Comment (efiop) [inline dvc/state.py]: Maybe we can retrieve it with sqlite3_limit ?

File: dvc/utils/__init__.py [hunk 6]
```diff
@@ -183,16 +183,26 @@ def remove(path):
...
 
-def to_chunks(l, jobs):
-    n = int(math.ceil(len(l) / jobs))
+def to_chunks(l, num_chunks=None, chunk_size=None):
+    if (num_chunks and chunk_size) or (not num_chunks and not chunk_size):
+        from dvc.exceptions import DvcException
 
-    if len(l) == 1:
-        return [l]
+        raise DvcException(
```
Comment (efiop) [inline dvc/utils/__init__.py]: Probably the `ValueError` is more appropriate here. Would unit test it though :)

File: dvc/utils/__init__.py [hunk 7]
```diff
@@ -183,16 +183,26 @@ def remove(path):
...
+            "One and only one of 'num_chunks', 'chunk_size' " "must be defined"
+        )
+
+    if num_chunks:
```
Comment (efiop) [inline dvc/utils/__init__.py]: Let's go the other way around, so we have less code in the branch. Something like:
```
if chunk_size:
    return [l[i : i + chunk_size] for i in range(0, len(l), chunk_size)]

# and now process `num_chunks` case
```

File: dvc/state.py (hunk 5 above)
Comment (pared) [reply dvc/state.py]: According to [docs](https://www.sqlite.org/c3ref/limit.html) `sqlite3_limit` is for setting the limit. Also I cannot find this method in python libs.

Comment (efiop) [reply dvc/state.py]: > Regardless of whether or not the limit was changed, the sqlite3_limit() interface returns the prior value of the limit.

File: dvc/state.py [hunk 8]
```diff
@@ -449,6 +450,10 @@ def remove_unused_links(self, used):
                 remove(path)
                 unused.append(relpath)
 
-        for relpath in unused:
-            cmd = "DELETE FROM {} WHERE path = ?".format(self.LINK_STATE_TABLE)
-            self._execute(cmd, (relpath,))
+        for chunk_unused in to_chunks(
+            unused, chunk_size=SQLITE_MAX_VARIABLES_NUMBER
+        ):
+            cmd = "DELETE FROM {} WHERE path IN ({})".format(
```
Comment (efiop) [inline dvc/state.py]: So MAX_PATH is ~4096 usually. so MAX_PATH * SQLITE_MAX_VARIABLES_NUMBER ~ 4M, and SQLITE_MAX_SQL_LENGTH ~ 1M. Theoretically, there might be a problem. But those numbers of files are usually in directories, for which we only add one entry - path to the directory itself, so it is unlikely that we will hit this, but if we do then sql will raise SQLITE_TOOBIG, and we can deal with it later.

**Comment by efiop** [comment]
@pared tests failed

File: dvc/utils/__init__.py [hunk 9]
```diff
@@ -183,8 +183,15 @@ def remove(path):
...
 
-def to_chunks(l, jobs):
-    n = int(math.ceil(len(l) / jobs))
+def to_chunks(l, num_chunks=None, chunk_size=None):
+    if (num_chunks and chunk_size) or (not num_chunks and not chunk_size):
+        raise ValueError(
+            "One and only one of 'num_chunks', 'chunk_size' " "must be defined"
+        )
+    if chunk_size:
+        return [l[i : i + chunk_size] for i in range(0, len(l), chunk_size)]
```
Comment (efiop) [inline dvc/utils/__init__.py]: Btw, this is the same line of code that is down below.

**Pull request merged**